lib/spack/spack/environment/environment.py

```diff
@@ -1024,6 +1024,47 @@ def add(self, user_spec, list_name=user_speclist_name):
 
         return bool(not existing)
 
+    def change_existing_spec(self, change_spec, list_name=user_speclist_name, match_spec=None):
+        """
+        Provide a single spec and assume you want to find another spec with the
+        same name and override it with any conflicting details in change_spec.
+
+        This is only allowed for environments that do not use matrices (i.e.
+        they have a single, explicitly-enumerated list of specs called
+        'specs').
```

Comment:
I think this should work on any list that doesn't contain a matrix, right?  Or, could it not be made to work on any single spec in a list, even if the list has matrices?

Reply:
Whoops this is out of date:

> I think this should work on any list that doesn't contain a matrix, right?

yes (and it does, and the tests check that).

> Or, could it not be made to work on any single spec in a list, even if the list has matrices?

possibly: the `SpecList` interface presents itself as a list so is opaque to single-element modifications. If it were refactored, this would be possible. IMO this PR is not the best place to do that.

Reply:
The docstring is now updated